contracts/deployer/AccessControlledDeployer.sol

```diff
@@ -0,0 +1,165 @@
+        if (admin == address(0) || pauser == address(0) || unpauser == address(0)) {
+            revert ZeroAddress();
+        }
+
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _grantRole(PAUSER_ROLE, pauser);
+        _grantRole(UNPAUSER_ROLE, unpauser);
+    }
+
+    /**
+     * @notice Deploys a contract using a deployment method defined by `deployer`
+     * @param deployer The create2 or create3 deployer contract that will deploy the contract
+     * @param bytecode The bytecode of the contract to be deployed
+     * @param salt A salt to influence the contract address
+     * @dev Only address with DEPLOYER_ROLE can call this function
+     * @dev This function requires that the current owner of `deployer` is this contract
+     * @dev The function can only be called if the contract is not in a paused state
+     * @dev The function emits `Deployed` event after the contract is deployed
+     * @return The address of the deployed contract
+     */
+    function deploy(
+        IDeployer deployer,
+        bytes memory bytecode,
+        bytes32 salt
+    ) external payable whenNotPaused onlyRole(DEPLOYER_ROLE) returns (address) {
+        if (address(deployer) == address(0)) {
+            revert ZeroAddress();
+        }
+        return deployer.deploy{value: msg.value}(bytecode, salt);
+    }
+
+    /**
+     * @notice Deploys a contract using a deployment method defined by `deployer` and initializes it
+     * @param deployer The create2 or create3 deployer contract that will deploy the contract
+     * @param bytecode The bytecode of the contract to be deployed
+     * @param salt A salt to influence the contract address
+     * @param init Init data used to initialize the deployed contract
+     * @dev Only address with DEPLOYER_ROLE can call this function
+     * @dev This function requires that the current owner of `deployer` is this contract
+     * @dev The function can only be called if the contract is not in a paused state
+     * @dev The function emits `Deployed` event after the contract is deployed
+     * @return The address of the deployed contract
+     */
+    function deployAndInit(
+        IDeployer deployer,
+        bytes memory bytecode,
+        bytes32 salt,
+        bytes calldata init
+    ) external payable whenNotPaused onlyRole(DEPLOYER_ROLE) returns (address) {
+        if (address(deployer) == address(0)) {
+            revert ZeroAddress();
+        }
+        return deployer.deployAndInit{value: msg.value}(bytecode, salt, init);
+    }
+
+    /**
+     * @notice Grants a list of addresses the DEPLOYER_ROLE
+     * @param deployers list of addresses to grant the DEPLOYER_ROLE
+     * @dev Only address with DEFAULT_ADMIN_ROLE can call this function
+     * @dev The function emits `RoleGranted` event for each address granted the DEPLOYER_ROLE.
+     *      This is not emitted if an address is already a deployer
+     */
+    function grantDeployerRole(address[] memory deployers) public {
```

Comment:
Any reason on not restricting this function call to when not paused? Same as the methods below.

Reply:
I think we'd want to enable admin operations even when the contract is paused, because this could be the way in which an admin resolves the issues that necessitated the pause. For instance if we pause the contract because a key with the DEPLOYER_ROLE is compromised, an admin would need to have the ability to revoke its permissions, before unpausing the bridge.